tests/unit/classifier/Perceptron_unittest.cc

```diff
@@ -68,16 +72,69 @@ TEST(Perceptron, custom_hyperplane_initialization)
 	auto test_features = wrap(env->get_features_test());
 	auto test_labels = wrap(env->get_labels_test());
 
-	auto perceptron = some<CPerceptron>(features, labels);
-	perceptron->train();
+	auto perceptron = some<CPerceptron>();
+	perceptron->set_labels(labels);
+	perceptron->train(features);
 
 	auto weights = perceptron->get_w();
 
-	auto perceptron_initialized = some<CPerceptron>(features, labels);
+	auto perceptron_initialized = some<CPerceptron>();
 	perceptron_initialized->set_initialize_hyperplane(false);
 	perceptron_initialized->set_w(weights);
-	perceptron_initialized->set_max_iter(1);
+	perceptron_initialized->put<int32_t>("max_iterations", 1);
+	perceptron_initialized->set_labels(labels);
 
-	perceptron_initialized->train();
+	perceptron_initialized->train(features);
 	EXPECT_TRUE(perceptron_initialized->get_w().equals(weights));
 }
+
+TEST(Perceptron, prematurely_stop)
+{
+	auto env = linear_test_env->getBinaryLabelData();
+	auto features = wrap(env->get_features_train());
+	auto labels = wrap(env->get_labels_train());
+	auto test_features = wrap(env->get_features_test());
+	auto test_labels = wrap(env->get_labels_test());
+
+	auto perceptron = some<CPerceptron>();
+	perceptron->set_labels(labels);
+	perceptron->train(features);
+
+	auto results = wrap(perceptron->apply(test_features))
```

Comment:
I think in c++, we can do this a bit simpler
`auto results = perceptron->apply(test_featatures)->as<CBinaryLabels>()->get_labels()`

Reply:
btw `wrap` is a hack we need for the meta examples, but not need to use it in c++